File: OTEAServer/Controllers/AmbitsController.cs
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;

namespace OTEAServer.Controllers
{
    /// <summary>
    /// Controller class for ambit operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>


    [ApiController]
    [Route("Ambits")]
    public class AmbitsController:ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context"></param>
        public AmbitsController(DatabaseContext context)
        {
            _context = context;
        }

        /// <summary>
        ///  Method that obtains from the database all the ambits
        /// </summary>
        /// <returns>All ambits</returns>
        [HttpGet("all")]
        public IActionResult GetAll()
        {
            try
            {
                var ambits = _context.Ambits.ToList();
                return Ok(ambits);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains from the database a ambit using its identifier
        /// </summary>
        /// <param name="id">Ambit identifier</param>
        /// <returns>The ambit if exists, null if not.</returns>

        [HttpGet("get")]
        public ActionResult<Ambit> Get([FromQuery] int id)
        {
            try
            {
                var ambit = _context.Ambits.FirstOrDefault(a => a.idAmbit == id);

                if (ambit == null)
                    return NotFound();

                return ambit;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that appends a new ambit to the database
        /// </summary>
        /// <param name="ambit">Ambit to append</param>
        /// <returns>Ambit append, null if not</returns>
        [HttpPost]
        public IActionResult Create([FromBody] Ambit ambit)
        {
            try
            {
                _context.Ambits.Add(ambit);
                _context.SaveChanges();
                return CreatedAtAction(nameof(Get), new { idAmbit = ambit.idAmbit }, ambit);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }


        /// <summary>
        /// 
        /// </summary>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="ambit">New ambit</param>
        /// <returns>Ambit if was updated, null if not</returns>
        [HttpPut]
        public IActionResult Update([FromQuery] int idAmbit, [FromBody] Ambit ambit)
        {
            try
            {
                if (idAmbit != ambit.idAmbit)
                    return BadRequest();

                var existingAmbit = _context.Ambits.FirstOrDefault(a => a.idAmbit == idAmbit);
                if (existingAmbit is null)
                    return NotFound();


                existingAmbit.idAmbit = ambit.idAmbit;
                existingAmbit.descriptionEnglish = ambit.descriptionEnglish;
                existingAmbit.descriptionSpanish = ambit.descriptionSpanish;
                existingAmbit.descriptionFrench = ambit.descriptionFrench;
                existingAmbit.descriptionBasque = ambit.descriptionBasque;
                existingAmbit.descriptionCatalan = ambit.descriptionCatalan;
                existingAmbit.descriptionDutch = ambit.descriptionDutch;
                existingAmbit.descriptionGalician = ambit.descriptionGalician;
                existingAmbit.descriptionGerman = ambit.descriptionGerman;
                existingAmbit.descriptionItalian = ambit.descriptionItalian;
                existingAmbit.descriptionPortuguese = ambit.descriptionPortuguese;
                _context.SaveChanges();

                return Ok(existingAmbit);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that deletes the ambit using its identifier
        /// </summary>
        /// <param name="id">Ambit identifier</param>
        /// <returns>Deleted ambit if deletion was succesful, null if not</returns>
        [HttpDelete]
        public IActionResult Delete([FromQuery] int id)
        {
            try
            {
                var ambit = _context.Ambits.FirstOrDefault(a => a.idAmbit == id);

                if (ambit is null)
                    return NotFound();

                _context.Ambits.Remove(ambit);
                _context.SaveChanges();

                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }
    }
}

File: connection/src/main/java/otea/connection/controller/AmbitsController.java
package otea.connection.controller;


import java.io.IOException;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.indicators.Ambit;
import otea.connection.ConnectionClient;
import otea.connection.api.AmbitsApi;
import retrofit2.Call;
import retrofit2.Response;
import session.Session;

/**
 * Controller class for ambit operations
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class AmbitsController {

    /**Ambits api to connect to the server*/
    private static AmbitsApi api;

    /**AmbitsController instance for singleton*/
    private static AmbitsController instance;

    private static int numAttempts=0;

    /**Class constructor*/
    private AmbitsController(){
        api= ConnectionClient.getInstance().getRetrofit().create(AmbitsApi.class);
    }

    /**
     * Synchronized method that obtains the instance
     *
     * @return Instance of ambit
     * */
    public static AmbitsController getInstance(){
        if(instance==null){
            synchronized (AmbitsController.class){
                if(instance==null){
                    instance=new AmbitsController();
                }
            }
        }
        return instance;
    }


    /**
     * Method that obtains from the database all the ambits
     *
     * @return All ambits
     * */
    public static List<Ambit> GetAll(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<Ambit>> callable = new Callable<List<Ambit>>() {
            @Override
            public List<Ambit> call() throws Exception {
                Call<List<Ambit>> call = api.GetAll(Session.getInstance().getToken());
                Response<List<Ambit>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<Ambit>> future = executor.submit(callable);
            List<Ambit> list = future.get();
            executor.shutdown();
            numAttempts=0;
            return list;
        } catch (InterruptedException | ExecutionException e) {
            if(e.getCause() instanceof SocketTimeoutException){
                numAttempts++;
                if(numAttempts<3) {
                    return GetAll();
                }
                else{
                    numAttempts=0;
                    return null;
                }
            }
            else{
                throw new RuntimeException(e);
            }
        }
    }

    /**
     * Method that obtains from the database a ambit using its identifier
     *
     * @param id - Ambit identifier
     * @return The ambit if exists, null if not.
     * */

    public static Ambit Get(int id){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Ambit> callable = new Callable<Ambit>() {
            @Override
            public Ambit call() throws Exception {
                Call<Ambit> call = api.Get(id,Session.getInstance().getToken());
                Response<Ambit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Ambit> future = executor.submit(callable);
            Ambit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that appends a new ambit to the database
     *
     * @param ambit - Ambit to append
     * @return Ambit append, null if not
     * */
    public static Ambit Create(Ambit ambit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Ambit> callable = new Callable<Ambit>() {
            @Override
            public Ambit call() throws Exception {
                Call<Ambit> call = api.Create(ambit,Session.getInstance().getToken());
                Response<Ambit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Ambit> future = executor.submit(callable);
            Ambit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that updates the ambit by the given
     *
     * @param idAmbit - Ambit identifier
     * @param ambit - New ambit
     * @return ambit if was updated, null if not
     *
     * */
    public static Ambit Update(int idAmbit, Ambit ambit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Ambit> callable = new Callable<Ambit>() {
            @Override
            public Ambit call() throws Exception {
                Call<Ambit> call = api.Update(idAmbit,ambit,Session.getInstance().getToken());
                Response<Ambit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Ambit> future = executor.submit(callable);
            Ambit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that deletes the ambit using its identifier
     *
     * @param id - Ambit identifier
     * @return Deleted ambit if deletion was succesful, null if not
     * */
    public static Ambit Delete(int id){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Ambit> callable = new Callable<Ambit>() {
            @Override
            public Ambit call() throws Exception {
                Call<Ambit> call = api.Delete(id,Session.getInstance().getToken());
                Response<Ambit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Ambit> future = executor.submit(callable);
            Ambit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }
}